### PR TITLE
Use a streamlined way to trigger component dismiss

### PR DESCRIPTION
**Pull request by GeoSot**
#### Till this MR:
 * plugins that support **dismiss** action by clicking on a `data-bs-dismiss` trigger, can be dismissed **ONLY** if the trigger is inside the plugin html.

* **Alert**, makes the difference.  Alert on  `data-bs-dismiss`  **click**,  tries to parse a `data-bs-target` element and in case in null, it fallback on the standard way, that assumes the trigger is inside the plugin html

#### This MR:
introduces the `enableDismissTrigger` helper function, that streamlines the functionality, using the idea of Alert component.

When 'dismiss' is clicked, it searches for `getElementFromSelector(event.target) || this.closest(`.${Plugin.NAME}`)` .

**Result:**

1. The dismiss trigger can be either inside the component, or with the use of `data-bs-target` it can be  anywhere in doc
2.  the dismiss trigger will **respect** the **disable** state and will handle **preventDefault** in case it is an  anchor

The idea came from #33324

### CODE REVIEW

**Comment by alpadev** [comment]
How about you define `enableDismissTrigger` on the base-component? That way you should be able to use it like `Plugin.enableDismissTrigger(callback)`.

**Review by alpadev — COMMENTED** [review]
How about defining `enableDismissTrigger` on the base-component? That way you could use it like `Plugin.enableDismissTrigger(callback)` and don't have to pass the plugin.

**Comment by GeoSot** [comment]
I don't know if it is ok to use it as static, as it will leave open the possibility to be called outside the code, causing fault usages

File: js/src/util/component-functions.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { getElementFromSelector, isDisabled } from './index'
+
+/**
+ * --------------------------------------------------------------------------
+ * Bootstrap (v5.0.1): util/component-functions.js
+ * Licensed under MIT (https://github.com/twbs/bootstrap/blob/main/LICENSE)
+ * --------------------------------------------------------------------------
+ */
+
+const enableDismissTrigger = (callback, Plugin) => {
```
Comment (alpadev) [inline js/src/util/component-functions.js:11]: IMO `Plugin` should go first and be `Component`.

**Review by alpadev — COMMENTED** [review]
https://github.com/twbs/bootstrap/blob/34218dab70d07a0421e4492620329011aafdabb6/js/src/modal.js#L144-L146

do we still need this here?

**Review by alpadev — COMMENTED** [review]
https://github.com/twbs/bootstrap/blob/34218dab70d07a0421e4492620329011aafdabb6/js/src/alert.js#L49

https://github.com/twbs/bootstrap/blob/34218dab70d07a0421e4492620329011aafdabb6/js/src/alert.js#L61-L63

This is covered by `enableDismissTrigger` too, isn't it?

**Comment by GeoSot** [comment]
> https://github.com/twbs/bootstrap/blob/34218dab70d07a0421e4492620329011aafdabb6/js/src/modal.js#L144-L146
> do we still need this here?

As we still support anchors with `href` triggers, I think yes. 



<br>


> https://github.com/twbs/bootstrap/blob/34218dab70d07a0421e4492620329011aafdabb6/js/src/alert.js#L49
> 
> https://github.com/twbs/bootstrap/blob/34218dab70d07a0421e4492620329011aafdabb6/js/src/alert.js#L61-L63
> 
> This is covered by `enableDismissTrigger` too, isn't it?

Is better to merge #33402 before this MR, because existing alert code, does some 'magic',  and makes me nervous, using other code

After #33402 is merged, we won't have to solve this riddle

**Comment by alpadev** [comment]
> > https://github.com/twbs/bootstrap/blob/34218dab70d07a0421e4492620329011aafdabb6/js/src/modal.js#L144-L146
> > 
> > do we still need this here?
> 
> As we still support anchors with `href` triggers, I think yes.

But this is done by `enableDismissTrigger` already?

**Comment by tungphan995** [comment]
coy

**Review by ffoodd — COMMENTED** [review]
As said in #33324, I'm a bit concerned about usability and accessibility. The refactor looks great, but use cases that it'd allow feel wrong.

Modal (and offcanvas) are supposed to make everything outside them inert while they're opened. Toast and alerts are IMHO valid use cases, but I'm not sure about this as a feature.

**Review by ffoodd — APPROVED** [review]
As said in #33403, this might need technical docs. Depending on this, we may need to add a callout regarding accessibility risks, but I guess we're pretty safe as long as we don't promote weird use cases.

File: site/content/docs/5.0/components/modal.md [hunk 2]
```diff
@@ -840,19 +840,10 @@ Activate a modal without writing JavaScript. Set `data-bs-toggle="modal"` on a c
...
-```
-
-or on a button **outside the modal** using the `data-bs-target` as demonstrated below:
-
-```html
-<button type="button" class="btn-close" data-bs-dismiss="modal" data-bs-target="#my-modal" aria-label="Close"></button>
-```
 {{< callout warning >}}
-While both ways to dismiss a modal are supported, keep in mind that dismissing from outside a modal does not match [the WAI-ARIA modal dialog design pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal). Do this at your own risk.
+While both ways to dismiss a modal, are supported, keep in mind that dismissing from outside a modal does not match [the WAI-ARIA modal dialog design pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal). Do this at your own risk.
```
Comment (ffoodd) [inline site/content/docs/5.0/components/modal.md:846]: This comma is unneeded :)

**Review by mdo — APPROVED** [review]
Love the consolidation of code and docs!

**Pull request merged**